File: multi-agent-ui/src/api/system/modelConfig.ts
import request, { type ApiResult } from "@/utils/request";
import { MODEL_CONFIG_URL } from "./index";

/** 模型配置实体（对应后端 sys_model_config.to_dict；api_key 为掩码值） */
export interface ModelConfig {
  id: string;
  /** 模型角色：chat / rewrite / visual / rerank，全局唯一，创建后不可变 */
  role: string;
  /** 卡片标题（中文名），如 '对话模型' */
  name: string;
  model_name: string;
  api_url: string;
  /** 掩码后的密钥（前 4 后 4，中间 ****）；编辑提交时留空则不修改 */
  api_key: string;
  /** provider 标识，目前仅 rerank 使用 */
  provider: string | null;
  /** 单请求超时（秒），目前 chat/visual 使用 */
  timeout: number | null;
  /** 失败重试次数，目前 chat/visual 使用 */
  max_retries: number | null;
  /** 角色特有参数兜底 */
  extra: Record<string, unknown> | null;
  /** 内置行删除保护标记，创建后不可变 */
  is_builtin: boolean;
  remark: string | null;
  created_at: string;
  updated_at: string;
}

/** 更新模型配置请求体（role/is_builtin 不可变；api_key 留空或为掩码占位则不修改） */
export interface ModelConfigUpdatePayload {
  name?: string;
  model_name?: string;
  api_url?: string;
  /** 仅在需要更换密钥时提交完整明文；留空则保持原值 */
  api_key?: string;
  provider?: string | null;
  timeout?: number | null;
  max_retries?: number | null;
  extra?: Record<string, unknown> | null;
  remark?: string | null;
}

/** 更新返回体：在模型配置基础上附带热更新是否生效标识 */
export type ModelConfigUpdateResult = ModelConfig & { refreshed?: boolean };

/** 模型配置管理 API：统一通过 modelConfigApi.xx() 调用 */
export const modelConfigApi = {
  /** 全量列出模型配置（api_key 掩码），供卡片页渲染 */
  list() {
    return request.get<ModelConfig[], ApiResult<ModelConfig[]>>(MODEL_CONFIG_URL.root);
  },

  /** 按角色精确查询单个模型配置（api_key 掩码） */
  getByRole(role: string) {
    return request.get<ModelConfig, ApiResult<ModelConfig>>(MODEL_CONFIG_URL.byRole(role));
  },

  /** 原子更新单行模型配置；成功后端触发配置快照刷新，结果经 data.refreshed 透出 */
  update(configId: string, payload: ModelConfigUpdatePayload) {
    return request.put<ModelConfigUpdateResult, ApiResult<ModelConfigUpdateResult>>(
      MODEL_CONFIG_URL.byId(configId),
      payload
    );
  },
};

File: entity/system/model_config.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, Integer, String, Text, func, text
from sqlalchemy.dialects.postgresql import JSONB, UUID
from sqlalchemy.orm import Mapped, mapped_column

from entity.base import Base
from utils.date_format import format_datetime
from utils.id import format_id


class ModelConfig(Base):
    """
    模型配置 ORM 模型（映射 sys_model_config）。

    - 每行对应一个模型角色（chat/rewrite/visual/rerank），role 全局唯一，
      前端以一行渲染为一张配置卡片；
    - api_key 落盘明文，掩码仅在 service 层做（to_dict 返回原值，禁止直接对外）；
    - provider 目前仅 rerank 使用；timeout/max_retries 目前 chat/visual 使用；
      extra 为角色特有参数兜底；
    - is_builtin 为内置行删除保护标记，禁删但允许更新其余字段；
    - updated_at 由业务层在更新时显式赋值，不依赖数据库触发器。
    """

    __tablename__ = "sys_model_config"
    __table_args__ = {"schema": "sys"}

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        server_default=text("uuidv7()"),
    )
    role: Mapped[str] = mapped_column(String(20), nullable=False)
    name: Mapped[str] = mapped_column(String(100), nullable=False)
    model_name: Mapped[str] = mapped_column(String(200), nullable=False)
    api_url: Mapped[str] = mapped_column(Text, nullable=False)
    api_key: Mapped[str] = mapped_column(Text, nullable=False)
    provider: Mapped[str | None] = mapped_column(String(50), nullable=True)
    timeout: Mapped[int | None] = mapped_column(Integer, nullable=True)
    max_retries: Mapped[int | None] = mapped_column(Integer, nullable=True)
    context_window: Mapped[int] = mapped_column(
        Integer, nullable=False, server_default=text("200000")
    )
    extra: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    is_builtin: Mapped[bool] = mapped_column(
        Boolean, nullable=False, server_default=text("TRUE")
    )
    remark: Mapped[str | None] = mapped_column(Text, nullable=True)

    created_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        nullable=False, server_default=func.now()
    )

    def to_dict(self) -> dict:
        """转为可 JSON 序列化的普通字典（api_key 为原值，掩码由 service 层处理）。"""
        return {
            "id": format_id(self.id),
            "role": self.role,
            "name": self.name,
            "model_name": self.model_name,
            "api_url": self.api_url,
            "api_key": self.api_key,
            "provider": self.provider,
            "timeout": self.timeout,
            "max_retries": self.max_retries,
            "context_window": self.context_window,
            "extra": self.extra,
            "is_builtin": self.is_builtin,
            "remark": self.remark,
            "created_at": format_datetime(self.created_at),
            "updated_at": format_datetime(self.updated_at),
        }
